websocket over http2

arenevier opened the pull request:
@ricea @jgraham

Commit:
Update to more recent versions of hyper libraries

Chrome wants to enable support for websocket over h2. In order to add
that support to wpt http2 server, we first need to update to update
hyper to a version that supports SETTINGS_ENABLE_CONNECT_PROTOCOL.

That support was added in January 2019, so we need versions at least as
recent. Since python 2.7 support was dropped in September 2020, we can't
take the very last versions. So we are taking the latest stable versions
before September.

Those are version 5.2.0 for hyperframe and version 3.2.0 for hyper-h2.
For hpack, that would be version 3.0.0, which is already the current
version.

arenevier commented:
This is the rfc for that feature https://github.com/web-platform-tests/rfcs/pull/67

Hexcles commented on tools/wptserve/wptserve/server.py:
```diff
@@ -641,7 +771,7 @@ class WebTestHttpd(object):
                  server_cls=None, handler_cls=Http1WebTestRequestHandler,
                  use_ssl=False, key_file=None, certificate=None, encrypt_after_connect=False,
-                 router_cls=Router, doc_root=os.curdir, routes=None,
+                 router_cls=Router, doc_root=os.curdir, ws_doc_root=os.curdir, routes=None,
```
The default for `ws_doc_root` should be `doc_root/websockets/handlers`

Hexcles commented on tools/wptserve/wptserve/server.py:
```diff
@@ -406,20 +424,125 @@ def handle_one_request(self):
         t.start()
         return t
 
+
```
Nit: drop an empty line here.
```suggestion
```

Hexcles commented on tools/wptserve/wptserve/server.py:
```diff
@@ -406,20 +424,125 @@ def handle_one_request(self):
+            except StreamClosedError: # maybe the stream has already been closed
+                pass
+        queue.put(None)
+
```
same here

Hexcles commented on tools/wptserve/wptserve/ws_h2_handshake.py:
```diff
@@ -0,0 +1,253 @@
+from six.moves import map
+from six.moves import range
+
+# TODO: We are using "private" methods of pywebsocket. We might want to
```
At least the exceptions are exposed in `mod_pywebsocket.handshake` via its `__init__.py`. We could do the same for other methods, too.

arenevier replied:
Any thought about that @ricea ?

ricea replied:
I'd like to do the refactor of pywebsocket after we land this, so we know exactly what we need to export, and can work out the nicest way to export it.

Hexcles commented on tools/wptserve/wptserve/server.py:
```diff
@@ -727,3 +858,55 @@ def get_url(self, path="/", query=None, fragment=None):
+
+    def read(self, length):
+        return self._request_handler.rfile.read(length)
+
```
Nit: two empty lines between top-level stuff.

Hexcles commented on tools/wptserve/wptserve/server.py:
```diff
@@ -511,6 +634,12 @@ def __init__(self, headers):
             dict.__setitem__(self, self._convert_h2_header_to_h1(key), val)
 
+    def get(self, key, def_value=None):
+        # pywebsocket will read the upper case value of those headers
```
This seems a bit brittle. Are headers in HTTP/2 also case-insensitive? If so, can we do case-insensitive lookup for all headers similar to what you can find in `request.py`? If not, this seems like a bug elsewhere that needs to be fixed.

arenevier replied:
HTTP/2 headers are case sensitive. I will modify the handshake code to read the correct (lower case) headers instead.

Hexcles reviewed (commented):
I reviewed the second commit (WebSockets over HTTP/2 in wptserve) and it looks pretty clean to me (note that I know little about HTTP/2 or WebSockets). Let's wait for the RFC to pass.

arenevier commented:
Error on macOS is: "No such file or directory" when trying to launch the browser. So it doesn't seem related to this patch.

https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=57067&view=logs&j=af75b4de-b26f-5d4d-58e8-b41258076d44&t=052839da-7227-5391-fe2d-4b66ec169a0c

Hexcles commented:
The RFC has landed, so I'm reviewing the PR now.

There are a few lint/stylistic errors (https://community-tc.services.mozilla.com/tasks/Vw217bFWThSgrwEHQlxqLQ/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FVw217bFWThSgrwEHQlxqLQ%2Fruns%2F0%2Fartifacts%2Fpublic%252Flogs%252Flive.log#L432). Could you fix them?

Hexcles commented on websockets/constants.js:
```diff
@@ -6,6 +6,9 @@ var PORT_SSL = "{{ports[wss][0]}}";
 if (location.search == '?wss') {
   SCHEME_DOMAIN_PORT = 'wss://{{host}}:' + PORT_SSL;
+} else if (location.search == '?http2') {
+  var PORT_H2 = "{{ports[h2][0]}}";
```
Shall we define this at the top of the file for consistency?

arenevier replied:
done

Hexcles reviewed (commented):
Overall the integration pieces LGTM.

One more thing, let's add an HTTP/2 case to `infrastructure/server/wpt-server-websocket.sub.html`. This will prevent the server from regressing.

arenevier commented on tools/wptserve/wptserve/server.py:
```diff
@@ -406,20 +424,123 @@ def handle_one_request(self):
             for stream_id, (thread, queue) in stream_queues.items():
                 thread.join()
 
+    # TODO(arenevier): should that method be moved to utils?
```
extract_method_header is in h2 third party library. So on second thought, it seems simpler to run leave it here.

arenevier commented on tools/wptserve/wptserve/server.py:
```diff
@@ -406,20 +424,123 @@ def handle_one_request(self):
+        for key, value in frame.headers:
+            if key in (b':protocol', u':protocol'):
+                if not isinstance(value, bytes):
+                    protocol = value.encode('utf-8')
```
That sounds good. I have replaced with isomorphic_encode

arenevier commented on tools/wptserve/wptserve/server.py:
```diff
@@ -406,20 +424,123 @@ def handle_one_request(self):
+            args=(self, request_wrapper, stream_handler, queue)
+        )
+        t.start()
+
```
done

arenevier commented:
> There are a few lint/stylistic errors (https://community-tc.services.mozilla.com/tasks/Vw217bFWThSgrwEHQlxqLQ/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FVw217bFWThSgrwEHQlxqLQ%2Fruns%2F0%2Fartifacts%2Fpublic%252Flogs%252Flive.log#L432). Could you fix them?

done

> One more thing, let's add an HTTP/2 case to `infrastructure/server/wpt-server-websocket.sub.html`. This will prevent the server from regressing.

I couldn't add a test to wpt-server-websocket.sub.html because it served over http1. I have added a new file http2-websocket.sub.h2.any.js

This test will fail on most current browsers (except I think, Firefox). 

Should I create a ini file in metadata/ ? If so, I'm not quite sure how to fill it.

ricea commented on tools/wptserve/wptserve/server.py:
```diff
@@ -406,20 +424,118 @@ def handle_one_request(self):
+            try:
+                # TODO(arenevier): reset_stream is needed. If we use end_stream,
+                # the websocket will not close cleanly. Is it according to spec?
+                connection.reset_stream(stream_id)
```
@bencebeky Do you know the answer to this question?

bencebeky replied:
This is not according to the spec.  According to https://tools.ietf.org/html/rfc8441#section-5, closing the HTTP/2 stream is like closing the TCP connection for regular WebSockets.

I was not aware of this bug, and I did not find any relevant issue on the Chromium bugtracker or any TODO in the source code.  Alas I did not find a test for this behavior either.  I suggest you file a bug at https://crbug.com/new and add a link to that in this comment.

arenevier replied:
I have opened https://bugs.chromium.org/p/chromium/issues/detail?id=1151393

@bencebeky: since this is not according to the spec, and if chromium has a bug, maybe I should remove the TODO and use end_stream here?

bencebeky replied:
Hm actually that makes a lot of sense to me.  Also, thank you for filing the bug.

ricea reviewed (approved):
lgtm!

jgraham commented on tools/wptserve/wptserve/server.py:
```diff
@@ -727,3 +844,56 @@ def get_url(self, path="/", query=None, fragment=None):
+
+
+class _WebSocketConnection(object):
+    """Mimic mod_python mp_conn."""
```
The docstring setup here is very odd. I think it's better to have the class-level description in the `__init__` docstring rather than having a generic description of what constructing the class does there.

jgraham commented on tools/wptserve/wptserve/server.py:
```diff
@@ -641,8 +756,9 @@ class WebTestHttpd(object):
-                 router_cls=Router, doc_root=os.curdir, routes=None,
-                 rewriter_cls=RequestRewriter, bind_address=True, rewrites=None,
+                 router_cls=Router, doc_root=os.curdir,
+                 ws_doc_root=os.path.join(os.curdir, "websockets", "handlers"),
```
I feel like this should default to None so it has to be explicitly configured.

jgraham reviewed (commented):
> I couldn't add a test to wpt-server-websocket.sub.html because it served over http1. I have added a new file http2-websocket.sub.h2.any.js
>
> This test will fail on most current browsers (except I think, Firefox).
>
> Should I create a ini file in metadata/ ? If so, I'm not quite sure how to fill it.

In `infrastructure/metadata/infrastructure/server/http2-websocket.sub.h2.any.js.ini`, something like:

```
[http2-websocket.sub.h2.any.html]
  [WSS over h2]
    expected:
      if product != "firefox": FAIL


[http2-websocket.sub.h2.any.worker.html]
  [WSS over h2]
    expected:
      if product != "firefox": FAIL
```

But test this with `wpt run --metadata infrastructure/metadata <browser> infrastructure/server/http2-websocket.sub.h2.any.js`.

jgraham commented on tools/wptserve/wptserve/ws_h2_handshake.py:
```diff
@@ -0,0 +1,248 @@
+class WsH2Handshaker(object):
+    """Opening handshake processor for the WebSocket protocol (RFC 6455)."""
+    def __init__(self, request, dispatcher):
+        """Construct an instance.
```
Another docstring to fix here.

arenevier replied:
done

jgraham reviewed (approved):
I also haven't looked at the HTTP/2 parts closely, but I assume there's enough expertise that they are already covered.

arenevier commented:
I have pushed a new change to this patch:
Since I am adding a **http2** variant, I need the ManifestItem to be considered h2 if the query is equal to **http2**. Without that change, the http2 variants won't be served over h2.

https://github.com/arenevier/wpt/commit/4ffd1c3eea02d9a2c75c6faff32d08e2b3f3735d#diff-8f280b64b0700ab9b4b343adabc7ff4d5ea4b1f45cb6c4bd7f50a19b21ebdb8c

@ricea @jgraham : does that change look good to you?

ricea commented:
> @ricea @jgraham : does that change look good to you?

Yes, lgtm.

jgraham commented:
Hmm. I'm not very excited about adding a new kind of filename flag that depends on a query string. That's not something we've had before. I wonder if we should consider a way to make the variant mechanism itself encode additional flags per variant, more like the `any.js` setup.

Commit:
WebSockets over HTTP/2 in wptserve

When we detect an extended CONNECT method, we don't use a the handler
provided by the server router. Instead, we create a websocket
dispatcher.

H2Response and H2Request are wrapped into _WebSocketRequest and
_WebSocketConnection. Those objects has the same interfaces and
mod_python request and mp_conn. That way, they can be used by
pywebosocket.

The handshake for the bootstrap websocket method is implemented in
ws_h2_handshake.py

When expose the h2 port in constants.js. Test who source it can add a
"http2" variant, and will use then connect to the websocket over the
http2 connection.

stephenmcgruer commented:
@jgraham and I discussed this this morning, and feel that a reasonable compromise is to use a query string, but have it be an actual query parameter e.g. `?flags=h2&...`. There was then further discussion on IRC (https://w3.logbot.info/testing/20201202#c595167):

```
jgraham arno: So I discussed this with smcgruer_[EST] a bit and I think we agree that something like <meta name=variant content="?flags=h2"> would be a good comprimise i.e. we'd reserve the flags query parameter for this purpose
smcgruer_[EST] Could also do `wpt_flags` or something if we wanna namespace it
smcgruer_[EST] Sorry, I have an AI to reply to the RFC with that and have not got to it yet
jgraham smcgruer_[EST] is outsourcing his job to GPT-3
arno jgraham: so it would be flags=h2, and the js would check location.search === '?flags=h2'
arno I will do a patch if you think that's the best solution.
arno In the meantime, what do you think of pushing the current patch (minus the latest addition)? The wptserver part is self contained, so maybe it can be merged already.
jgraham arno: Yeah I think that allows other tests to use this without needing to be exactly like the websockets tests, or hardcoding a bunch of cases
jgraham I think it was approved before that addition? In which case I'm happy for you to push that part :)
arno great. I pushed a new patch.
```

arenevier commented:
I've added support wpt_flags in https://github.com/arenevier/wpt/commit/977e3ee799aaafc35ca08fc99a0e8dac49935fed

@jgraham @stephenmcgruer @ricea

jgraham commented:
That looks pretty good to me. I guess there's a question about whether one can pass multiple flags via the query parameter e.g. both `www` and `https`. @gsnedders do you have thoughts on this?

arenevier commented:
> That looks pretty good to me. I guess there's a question about whether one can pass multiple flags via the query parameter e.g. both `www` and `https`. @gsnedders do you have thoughts on this?

In the current implementation, thanks to how python parse_qs works, you can pass multiple wpt_flags parameters.

`wpt_flags=www&wpt_flags=https`

stephenmcgruer commented on tools/wptserve/wptserve/server.py:
```diff
@@ -406,20 +424,116 @@ def handle_one_request(self):
+
+        request_wrapper._dispatcher = dispatcher
+
+        # TODO(arenevier): we need two threads:
+        # - one to handle the frame queue
+        # - one to handle the request (dispatcher.transfer_data is blocking)
+        # the alternative is to have only one (blocking) thread. That thread
+        # will call transfer_data. That would require a special case in
+        # handle_one_request, to bypass the queue and write data to wfile
+        # directly.
```
It's not clear to me what this `TODO` is for. Is this something that should be fixed in the future (e.g. move to one thread)?

stephenmcgruer commented on tools/manifest/item.py:
```diff
@@ -120,6 +120,9 @@ def __init__(self,
         self._url = url
         self._extras = extras
+        parsed_url = urlparse(self.url)
+        self._flags = (set(parsed_url.path.rsplit("/", 1)[1].split(".")[1:-1])
+                       | set(parse_qs(parsed_url.query).get("wpt_flags", [])))
```
This is a lint error, from the unittests bot:

`./manifest/item.py:125:24: W503 line break before binary operator`

stephenmcgruer reviewed (approved):
LGTM modulo some nits and small comments. It may have been nice to add some unittests for the `tools/manifest` and `tools/wptserve` changes, given all we have currently is a pair of end-to-end infrastructure tests, but I won't hold up the CL over it.

Commit:
Add wpt_flags query param.

Test flags can be passed via the wpt_flags query parameter.

arenevier commented:
> LGTM modulo some nits and small comments. It may have been nice to add some unittests for the `tools/manifest` and `tools/wptserve` changes, given all we have currently is a pair of end-to-end infrastructure tests, but I won't hold up the CL over it.

I have added a unittest in tools/manifest. Adding unittests to tools/wptserve seems more tricky.
